src/eng_back.c

```diff
@@ -605,11 +644,45 @@ static void *ctx_load_object(ENGINE_CTX *ctx,
 /* Certificate handling                                                       */
 /******************************************************************************/
 
+static PKCS11_CERT *cert_cmp(PKCS11_CERT *a, PKCS11_CERT *b, time_t *ptime)
+{
+	const ASN1_TIME *aa, *ba;
+	int pday, psec;
+
+	/* the best certificate exists */
+	if (!a || !a->x509) {
+		return b;
+	}
+	if (!b || !b->x509) {
+		return a;
+	}
+
+	aa = X509_get0_notAfter(a->x509);
+	ba = X509_get0_notAfter(b->x509);
+
+	/* the best certificate expires last */
+	if (ASN1_TIME_diff(&pday, &psec, aa, ba)) {
+		if (pday < 0 || psec < 0) {
+			return a;
```

Comment:
If you check for the end of expiration date and select the later one, you should check also for the start of validity. In rare case, you might get certificate with validity starting in the future and this cmp function will select it.

Reply:
My original attempt to fix this took into account the start of validity, and then I realised to make a decision with respect to the start of validity you need the current time, but there was no concrete mechanism I could use to guarantee that I could get the correct time.

For low security stuff the system clock might be enough, for higher security stuff where the HSM has a clock on board getting access to that clock reliably was something I couldn't figure out. I didn't want any behaviour that wasn't 100% deterministic.

I eventually settled on expiry date only, as that comparison always yields the same winner. The cost of this is that people deploying certs shouldn't deploy certs that aren't valid yet, which seems to be a sensible limitation, and if you do need to distinguish in this case where two certs issued at different times have the same ID (potentially because same subject key identifier because same key), use a label.